### ISSUE TITLE
OutOfMemoryException while creating a snapshot

### ISSUE THREAD

**Issue by ryanlewis** [KevinJump/uSync3.0#41]
Admittedly, this is on a site with a lot of doctypes and templates.

On a 8GB PC and the site running in debug mode, I managed to get an OutOfMemoryException while trying to create a snapshot. The snapshot itself should have consisted of a composition doctype consisting of one tab and 5 properties, and changes to 2 doctypes adding the composition to its makeup. 

The error appears as a simple message within the uSync snapshot admin page.

I imagine this might come as a bit of a surprise, but if there's any diagnostic or logging steps you wish me to take. 

Running Umbraco 7.4.1:

packages.config:

```
  <package id="uSync" version="3.1.1.740" targetFramework="net452" />
  <package id="uSync.Core" version="5.3.1.740" targetFramework="net452" />
  <package id="uSync.Snapshots" version="1.0.2.740" targetFramework="net452" />
```

**Comment by KevinJump** [comment]
Hi Ryan

is there anything in the umbraco logs in `/app_data/logs/` ? 

the dashboard only really displays the reason, but usually the logs have more detail. 

Kevin

**Comment by ryanlewis** [comment]
Apologies, that really would have been useful for you first time round:

Actually observed two OOM exceptions in quick succession. Sample below:

```
2016-04-19 18:15:23,327 [10] INFO  umbraco.content - Save Xml to file...
2016-04-19 18:15:25,523 [10] ERROR umbraco.content - Failed to save Xml to file.
System.OutOfMemoryException: Exception of type 'System.OutOfMemoryException' was thrown.
   at System.Text.StringBuilder.ExpandByABlock(Int32 minBlockCharCount)
   at System.Text.StringBuilder.Append(Char* value, Int32 valueCount)
   at System.Text.StringBuilder.Append(Char[] value, Int32 startIndex, Int32 charCount)
   at System.IO.StringWriter.Write(Char[] buffer, Int32 index, Int32 count)
   at System.Xml.XmlEncodedRawTextWriter.FlushBuffer()
   at System.Xml.XmlEncodedRawTextWriter.WriteAttributeTextBlock(Char* pSrc, Char* pSrcEnd)
   at System.Xml.XmlEncodedRawTextWriter.WriteString(String text)
   at System.Xml.XmlEncodedRawTextWriterIndent.WriteString(String text)
   at System.Xml.XmlWellFormedWriter.WriteString(String text)
   at System.Xml.XmlText.WriteTo(XmlWriter w)
   at System.Xml.XmlAttribute.WriteContentTo(XmlWriter w)
   at System.Xml.XmlAttribute.WriteTo(XmlWriter w)
   at System.Xml.XmlElement.WriteStartElement(XmlWriter w)
   at System.Xml.XmlElement.WriteElementTo(XmlWriter writer, XmlElement e)
   at System.Xml.XmlElement.WriteTo(XmlWriter w)
   at System.Xml.XmlDocument.WriteContentTo(XmlWriter xw)
   at System.Xml.XmlDocument.WriteTo(XmlWriter w)
   at umbraco.content.SaveXmlToString(XmlDocument xml)
   at umbraco.content.<SaveXmlToFileAsync>d__b.MoveNext()
2016-04-19 18:15:25,759 [10] ERROR Umbraco.Core.UmbracoApplicationBase - An unhandled exception occurred
System.Web.HttpUnhandledException (0x80004005): Exception of type 'System.Web.HttpUnhandledException' was thrown. ---> System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.OutOfMemoryException: Exception of type 'System.OutOfMemoryException' was thrown.
   at System.String.CreateStringFromEncoding(Byte* bytes, Int32 byteLength, Encoding encoding)
   at System.Text.UnicodeEncoding.GetString(Byte[] bytes, Int32 index, Int32 count)
   at System.Data.SqlClient.TdsParserStateObject.TryReadString(Int32 length, String& value)
   at System.Data.SqlClient.TdsParser.TryReadSqlStringValue(SqlBuffer value, Byte type, Int32 length, Encoding encoding, Boolean isPlp, TdsParserStateObject stateObj)
   at System.Data.SqlClient.TdsParser.TryReadSqlValue(SqlBuffer value, SqlMetaDataPriv md, Int32 length, TdsParserStateObject stateObj, SqlCommandColumnEncryptionSetting columnEncryptionOverride, String columnName)
   at System.Data.SqlClient.SqlDataReader.TryReadColumnData()
   at System.Data.SqlClient.SqlDataReader.TryReadColumnInternal(Int32 i, Boolean readHeaderOnly)
   at System.Data.SqlClient.SqlDataReader.TryReadColumn(Int32 i, Boolean setTimeout, Boolean allowPartiallyReadColumn)
   at System.Data.SqlClient.SqlDataReader.GetString(Int32 i)
   at umbraco.DataLayer.RecordsReaderAdapter`1.GetString(String fieldName)
   at InfoCaster.Umbraco.UrlTracker.Repositories.UrlTrackerRepository.GetUrlTrackerEntries(Nullable`1 maximumRows, Nullable`1 startRowIndex, String sortExpression, Boolean _404, Boolean include410Gone, Boolean showAutoEntries, Boolean showCustomEntries, Boolean showRegexEntries, String keyword, Boolean onlyForcedRedirects) in D:\kipusoep\Documents\GitHub\UrlTracker\Repositories\UrlTrackerRepository.cs:line 240
   at InfoCaster.Umbraco.UrlTracker.Repositories.UrlTrackerRepository.GetNotFoundEntries(Nullable`1 maximumRows, Nullable`1 startRowIndex, String sortExpression, String keyword) in D:\kipusoep\Documents\GitHub\UrlTracker\Repositories\UrlTrackerRepository.cs:line 318
   at InfoCaster.Umbraco.UrlTracker.Repositories.UrlTrackerRepository.GetNotFoundEntries(String sortExpression) in D:\kipusoep\Documents\GitHub\UrlTracker\Repositories\UrlTrackerRepository.cs:line 400
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Web.UI.WebControls.ObjectDataSourceView.InvokeMethod(ObjectDataSourceMethod method, Boolean disposeInstance, Object& instance)
   at System.Web.UI.WebControls.ObjectDataSourceView.ExecuteSelect(DataSourceSelectArguments arguments)
   at System.Web.UI.DataSourceView.Select(DataSourceSelectArguments arguments, DataSourceViewSelectCallback callback)
   at System.Web.UI.WebControls.DataBoundControl.PerformSelect()
   at System.Web.UI.WebControls.BaseDataBoundControl.DataBind()
   at System.Web.UI.WebControls.GridView.DataBind()
   at System.Web.UI.WebControls.BaseDataBoundControl.EnsureDataBound()
   at System.Web.UI.WebControls.CompositeDataBoundControl.CreateChildControls()
   at System.Web.UI.Control.EnsureChildControls()
   at System.Web.UI.WebControls.CompositeDataBoundControl.get_Controls()
   at ClientDependency.Core.ControlExtensions.FlattenChildren(Control control) in x:\Projects\ClientDependency\ClientDependency\ClientDependency.Core\ControlExtensions.cs:line 16
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__58`1.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__58`1.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__58`1.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__58`1.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__58`1.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__58`1.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__58`1.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__58`1.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__58`1.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__58`1.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__58`1.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__58`1.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__58`1.MoveNext()
   at System.Linq.Enumerable.<SelectManyIterator>d__16`2.MoveNext()
   at System.Linq.Enumerable.<ConcatIterator>d__58`1.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at ClientDependency.Core.Controls.ClientDependencyLoader.FindDependencies(Control control) in x:\Projects\ClientDependency\ClientDependency\ClientDependency.Core\Controls\ClientDependencyLoader.cs:line 334
   at ClientDependency.Core.Controls.ClientDependencyLoader.RegisterClientDependencies(WebFormsFileRegistrationProvider provider, Control control, IEnumerable`1 paths) in x:\Projects\ClientDependency\ClientDependency\ClientDependency.Core\Controls\ClientDependencyLoader.cs:line 323
   at ClientDependency.Core.Controls.ClientDependencyLoader.OnPreRender(EventArgs e) in x:\Projects\ClientDependency\ClientDependency\ClientDependency.Core\Controls\ClientDependencyLoader.cs:line 137
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Control.PreRenderRecursiveInternal()
   at System.Web.UI.Page.ProcessRequestMain(Boolean includeStagesBeforeAsyncPoint, Boolean includeStagesAfterAsyncPoint)
   at System.Web.UI.Page.HandleError(Exception e)
   at System.Web.UI.Page.ProcessRequestMain(Boolean includeStagesBeforeAsyncPoint, Boolean includeStagesAfterAsyncPoint)
   at System.Web.UI.Page.ProcessRequest(Boolean includeStagesBeforeAsyncPoint, Boolean includeStagesAfterAsyncPoint)
   at System.Web.UI.Page.ProcessRequest()
   at System.Web.UI.Page.ProcessRequestWithNoAssert(HttpContext context)
   at System.Web.UI.Page.ProcessRequest(HttpContext context)
   at ASP.umbraco_urltracker_infocaster_umbraco_urltracker_ui_urltrackermanager_aspx.ProcessRequest(HttpContext context) in c:\Users\ryan.lewis\AppData\Local\Temp\Temporary ASP.NET Files\vs\94505cf3\672d55fe\App_Web_infocaster.umbraco.urltracker.ui.urltrackermanager.aspx.6b1c27ca.wvo4uqns.0.cs:line 0
   at System.Web.HttpApplication.CallHandlerExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute()
   at System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously)
```

What is interesting is that you see OOM coming from the Core at the same time. If uSync is not causing Umbraco to republish its XML cache, the distributed cache update instruction functionality in 7.3/7.4 could have meant that another developer caused my running version of Umbraco to update itself as a result of them publishing some work. This resulted in two intensive jobs occurring at the same time, and exhausting all the memory maybe?

**Comment by KevinJump** [comment]
yeah it's a bit odd :worried: 

if you are taking a snapshot, usync is just doing readonly things (so no publishing or saving inside umbraco). Its possible that something is leaking in usync and that has tripped some memory issues that it then looks like a publish or save has triggered. 

As a guess how many doctypes, datatypes, etc do you think your site has, i might just create a site with loads of doctypes and see if the memory does anything strange.

**Comment by ryanlewis** [comment]
We've managed to create an initial snapshot when first developing on this inherited site. We found the first snapshot resulted in 553 directories, 550 files total. We have also since done some additional snapshots since the initial one without issue (albeit slow). In total, we have 5 snapshots, the big one detailed above and 4 additional ones with reasonably minor changes (1-2 doctypes affected, 1-2 datatypes added/changed).
